.github/workflows/tests.yml

```diff
@@ -2,43 +2,67 @@ name: Tests
 
 on:
   push:
-    branches:
-      - master
+    branches: [master]
   pull_request:
 
 jobs:
   unit-tests:
+    strategy:
+      matrix:
+        os: [ 'ubuntu-latest' ]
+        ruby: [ '2.5' ]
+        bundler: [ '1.17.3' ]
+    runs-on: ${{ matrix.os }}
     name: Unit tests
-    runs-on: ubuntu-latest
 
     steps:
-      - name: Checkout
+      - name: Checkout ${{ github.sha	}}
         uses: actions/checkout@v2
-      - name: Setup Ruby
+
+      - name: Setup Ruby ${{ matrix.ruby }}
         uses: actions/setup-ruby@v1
         with:
-          ruby-version: '2.5'
+          ruby-version: ${{ matrix.ruby }}
+
       - name: Cache dependencies
         uses: actions/cache@v2
         with:
           path: ~/vendor/bundle
           key: ${{ runner.os }}-gems-${{ hashFiles('apps/*/Gemfile.lock') }}
-      - name: Setup Bundler
+          restore-keys: |
+            ${{ runner.os }}-gems-
+
+      - name: Setup Bundler ${{ matrix.bundler }}
         run: |
-          gem install bundler -v 1.17.3
+          gem install bundler -v ${{ matrix.bundler }}
           bundle config path ~/vendor/bundle
+
       - name: Run ShellCheck
         run: rake test:shellcheck
+
       - name: Run unit tests
         run: rake test:unit
-  
+
   e2e-tests:
+    strategy:
+      matrix:
+        os: [ 'ubuntu-latest' ]
+        ruby: [ '2.5' ]
+        bundler: [ '1.17.3' ]
+    runs-on: ${{ matrix.os }}
     name: End-to-end tests
-    runs-on: ubuntu-latest
-    needs: unit-tests
 
     steps:
-      - name: Checkout
+      - name: Checkout ${{ github.sha	}}
         uses: actions/checkout@v2
+
+      - name: Cache dependencies
+        uses: actions/cache@v2
+        with:
+          path: ~/vendor/bundle
+          key: ${{ runner.os }}-gems-${{ hashFiles('apps/*/Gemfile.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-gems-
```

Comment:
What's the cache being used for here? Everyting in the e2e tests are inside a container, so any new gems are being installed in _it_.